vault/logical_system_raft.go

```diff
@@ -236,26 +237,46 @@ func (b *SystemBackend) handleRaftRemovePeerUpdate() framework.OperationFunc {
 	}
 }
 
+func (b *SystemBackend) getRaftBootstrapChallengeAnswer(serverID string) []byte {
+	// We take our local key and the given serverID and perform an HKDF
+	// invocation on it. This output is then encrypted using the root key
+	// (which protects our barrier keyring) and the remote peer can
+	// "prove" they're allowed to join by unsealing the challenge answer
+	// and providing it back to us. We assume inverting HKDF is hard (it
+	// is at least as hard as inverting an HMAC) and it is hard for an
+	// attacker to find (serverID, answer) such that:
+	//
+	//     HKDF(challengeKey, serverID) == answer
+	//
+	// The benefit of this is that we only need constant memory (a single
+	// root challenge key) and HKDF is fast and arbitrarily extensible.
+
+	kdf := hkdf.New(sha256.New, b.Core.pendingRaftPeerChallengeKey, []byte("openbao-raft-peer-challenge"), []byte(serverID))
+
+	// hkdf cannot fail on this short of output. While this value was
+	// previously 16 bytes, the root key is 32 bytes so 24 bytes provides
+	// increased safety while still retaining domain safety from the root
+	// key.
+	var answer [24]byte
+	_, err := kdf.Read(answer[:])
+	if err != nil {
+		panic(fmt.Sprintf("hkdf failed on 24 bytes of output: %v", err))
```

Comment:
340% sure ;-)

We read exactly once, of 24 bytes, out of a maximum theoretical of `(counter size * hash size) = 255*32 = 8,160 bytes`.

 - https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.29.0:hkdf/hkdf.go;l=92
 - https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.29.0:hkdf/hkdf.go;l=86
 - https://cs.opensource.google/go/x/crypto/+/refs/tags/v0.29.0:hkdf/hkdf.go;l=49

The Go toolchain even panics rather than erring:

- https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/crypto/tls/key_schedule.go;l=66
- https://cs.opensource.google/go/go/+/refs/tags/go1.23.3:src/crypto/internal/hpke/hpke.go;l=48

Though, by same reasoning above, `|| n != int(length)` should be redundant and never hit except in the error case since we validate we can return all requested bytes in HKDF, I'm happy to add that.

I'm also happy to pull it out into an error if we prefer, though.

Reply:
340% is good enough for me :smile: 

Thanks for explaining!